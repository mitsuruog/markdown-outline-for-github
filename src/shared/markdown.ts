export function parse($markdown: Element) {
  const $ul = document.createElement("ul");
  const $outline = $markdown.querySelectorAll("h1, h2, h3, h4, h5, h6");
  Array.from($outline).forEach(($element) => {
    const title = $element.textContent;
    const tagName = $element.tagName.toLowerCase();
    const href = $element.querySelector("a").href;
    const $anchor = document.createElement("a");
    $anchor.href = href;
    $anchor.textContent = title;
    const $li = document.createElement("li");
    $li.classList.add(
      `mitsuruog-markdown-outline__${tagName}`,
      "mitsuruog-markdown-outline__item"
    );
    $li.appendChild($anchor);
    $ul.appendChild($li);
  });
  $ul.classList.add("mitsuruog-markdown-outline__item-wrapper");
  return $ul;
}

export function getMarkdown(pathname: string, $body: HTMLElement) {
  const contents = $body.querySelectorAll(
    "article.markdown-body, #wiki-body .markdown-body"
  );
  return contents.length > 0 ? contents[0] : null;
}
