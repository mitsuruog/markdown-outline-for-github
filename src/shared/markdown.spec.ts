import { getMarkdown, parse } from "./markdown";
import "@testing-library/jest-dom";

describe("Test getMarkdown", () => {
  test("should return null if the content is empty", () => {
    const container = document.createElement("div");
    expect(getMarkdown("", container)).toEqual(null);
  });
  test("should return null if the content does not match", () => {
    const container = document.createElement("div");
    container.innerHTML = `
      <article>
        <div>Markdown</div>
      </article>
    `;
    expect(getMarkdown("", container)).toEqual(null);
  });
  test("should return child nodes if the content matches", () => {
    const container = document.createElement("div");
    container.innerHTML = `
      <article class="markdown-body">
        <div>Markdown</div>
      </article>
    `;
    expect(getMarkdown("", container)).toHaveTextContent("Markdown");
  });
});

describe("Test parse", () => {
  let result: HTMLElement | null = null;
  beforeEach(() => {
    const container = document.createElement("div");
    container.innerHTML = `
      <h1><a href="/h1">H1</a></h1>
      <h2><a href="/h2">H2</a></h2>
      <h3><a href="/h3">H3</a></h3>
      <h4><a href="/h4">H4</a></h4>
      <h5><a href="/h5">H5</a></h5>
      <h6><a href="/h6">H6</a></h6>
    `;
    result = parse(container);
  });

  test("should parse the markdown", () => {
    expect(result).toMatchSnapshot();
  });

  test("should parse h1 tag", () => {
    const h1 = result.querySelector("mitsuruog-markdown-outline__h1");
    expect(h1).toBeDefined();
  });
});
