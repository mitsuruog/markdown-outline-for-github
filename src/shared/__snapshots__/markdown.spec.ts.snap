// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test parse should parse the markdown 1`] = `
<ul
  class="mitsuruog-markdown-outline__item-wrapper"
>
  <li
    class="mitsuruog-markdown-outline__h1 mitsuruog-markdown-outline__item"
  >
    <a
      href="http://localhost/h1"
    >
      H1
    </a>
  </li>
  <li
    class="mitsuruog-markdown-outline__h2 mitsuruog-markdown-outline__item"
  >
    <a
      href="http://localhost/h2"
    >
      H2
    </a>
  </li>
  <li
    class="mitsuruog-markdown-outline__h3 mitsuruog-markdown-outline__item"
  >
    <a
      href="http://localhost/h3"
    >
      H3
    </a>
  </li>
  <li
    class="mitsuruog-markdown-outline__h4 mitsuruog-markdown-outline__item"
  >
    <a
      href="http://localhost/h4"
    >
      H4
    </a>
  </li>
  <li
    class="mitsuruog-markdown-outline__h5 mitsuruog-markdown-outline__item"
  >
    <a
      href="http://localhost/h5"
    >
      H5
    </a>
  </li>
  <li
    class="mitsuruog-markdown-outline__h6 mitsuruog-markdown-outline__item"
  >
    <a
      href="http://localhost/h6"
    >
      H6
    </a>
  </li>
</ul>
`;
